File: Backtracking/1079_M_LetterTilePossibilities.js
/**
 * https://leetcode.com/problems/letter-tile-possibilities/
 * 5.20 night 5.21 night
 */

/**
 * https://www.acwing.com/solution/LeetCode/content/2468/
 * Accepted --- 76ms 34.6MB 67.32%
 */
let ans = 0;
const numTilePossibilities_acwing = (tiles) => {
    let seen = new Map();
    let num = [];
    for (const c of tiles) {
        if (seen[c] == seen[-1]) { // check if the iterator hit the end of Map
            // if (seen[c] == seen[seen.length - 1]) { 
            seen[c] = num.length;
            num.push(0);
        }
        num[seen[c]]++;
    }
    ans = 0;
    dfs(num);
    return ans - 1;
};

const dfs = (num) => {
    ans++;
    for (let i = 0; i < num.length; i++) {
        if (num[i] > 0) {
            num[i]--;
            dfs(num);
            num[i]++;
        }
    }
};

/**
 * https://www.cnblogs.com/hwd9654/p/11008562.html
 * Accepted --- 76ms 35.3MB 67.32%
 */
const numTilePossibilities_cnblogs = (tiles) => {
    if (tiles.length == 0) return 0;
    let count = [];
    fillArr(count, 26);
    for (const c of tiles.split("")) {
        let tmp = c.charCodeAt(0);
        count[tmp - 65]++; // c - 'A' will return NaN, need to transfer to ASCII manually
    }
    // console.log(count);
    return DFS(count);
}

const DFS = (array) => {
    let res = 0;
    for (let i = 0; i < 26; i++) {
        if (array[i] == 0) continue;
        res++;
        array[i]--;
        res += DFS(array);
        array[i]++;
    }
    return res;
}

const fillArr = (arr, n) => {
    for (let i = 0; i <= n; i++) {
        arr.push(0);
    }
};

// not fixed
const numTilePossibilities_csdn = (tiles) => {
    let count = [];
    fillArr(count, 25);
    for (const c of tiles) {
        // count[c - 'A']++;
        let tmp = c.charCodeAt(0);
        count[tmp - 65]++;
    }
    let res = 0;
    backtrack(count, res);
    console.log(count)
    console.log(count.length)
    return res;
};

const backtrack = (count, res) => {
    for (let i = 0; i < 26; i++) {
        if (count[i] == 0) {
            continue;
        }
        res++;
        console.log(res); // problem
        count[i]--;
        backtrack(count, res);
        count[i]++;
    }
    return res;
};

const main = () => {
    let tiles = "AAB";
    let tiles2 = "AAABBC";

    console.log(numTilePossibilities_acwing(tiles));
    console.log(numTilePossibilities_acwing(tiles2));

    console.log("")
    console.log(numTilePossibilities_cnblogs(tiles));
    console.log(numTilePossibilities_cnblogs(tiles2));

    // console.log("")
    // console.log(numTilePossibilities_csdn(tiles));
    // console.log(numTilePossibilities_csdn(tiles2));
};

main()

File: Backtracking/1079_M_LetterTilePossibilities.cxx
// 5.20 night 5.21 night
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    void print(vector<int> const &input)
    {
        for (int i = 0; i < input.size(); i++)
        {
            cout << input.at(i) << ' ';
        }
        cout << endl;
    }

    void print_map(unordered_map<char, int> map)
    {
        for (const auto &p : map)
            cout << "map[" << p.first << "] = " << p.second << '\n';
    }

    // https://blog.csdn.net/fuxuemingzhu/article/details/101354494
    int numTilePossibilities_csdn(string tiles)
    {
        vector<int> count(26, 0);
        for (char c : tiles)
        {
            count[c - 'A']++;
        }
        int res = 0;

        // print(count);
        // cout << count.size() << endl;
        backtrack(count, res);

        // print(count);
        // cout << count.size() << endl;
        return res;
    }
    void backtrack(vector<int> &count, int &res)
    {
        for (int i = 0; i < 26; ++i)
        {
            if (count[i] == 0)
            {
                continue;
            }
            res++;
            // cout << res << endl;
            count[i]--;
            backtrack(count, res);
            count[i]++;
        }
    }

    int ans;
    int numTilePossibilities_acwing(string tiles)
    {
        unordered_map<char, int> seen;
        vector<int> num;
        print_map(seen);
        for (auto &c : tiles)
        {
            if (seen.find(c) == seen.end())
            {
                cout << c << endl;
                // cout << seen[c] << endl;
                seen[c] = num.size();
                num.push_back(0);
            }
            num[seen[c]]++;
        }
        print(num);
        ans = 0;
        dfs(num);
        return ans - 1;
    }

    void dfs(vector<int> &num)
    {
        ans++;
        for (int i = 0; i < num.size(); i++)
        {
            if (num[i] > 0)
            {
                num[i]--;
                dfs(num);
                num[i]++;
            }
        }
    }
};

int main()
{
    Solution s;
    string tiles = "AAB";
    string tiles2 = "AAABBC";

    cout << s.numTilePossibilities_csdn(tiles) << endl;
    // cout << s.numTilePossibilities_csdn(tiles2) << endl;

    cout << endl
         << s.numTilePossibilities_acwing(tiles) << endl;
    // cout << s.numTilePossibilities_acwing(tiles2) << endl;
}